RFCs/0002-pluggable-discovery.md

```diff
@@ -0,0 +1,854 @@
+
+Let’s see how a complete example will look like:
+
+```
+tools.arduino_ota.upload.field.username=Username
+tools.arduino_ota.upload.field.password=Password
+tools.arduino_ota.upload.field.password.secret=true
+tools.arduino_ota.upload.pattern="{runtime.tools.arduinoOTA.path}/bin/arduinoOTA"
+                                 -address {upload.port.address} -port 65280
+                                 -username "{upload.field.username}
+                                 -password "{upload.field.password}"
+                                 -sketch "{build.path}/{build.project_name}.bin"
+```
+
+#### Support for upload "without a port" (`no-port` dummy protocol)
+
+Some upload tools already have the port detection builtin so there is no need to specify an upload port (for example the `openocd` tool is often able to autodetect the devices to upload by itself).
+
+To support this particular use case a dummy `no-port` protocol has been reserved:
+
+```
+myboard.upload.tool.no-port=openocd_without_port
+```
+
+The `no-port` upload recipe will be selected when:
+
+- The upload port is not specified
+
+or
+
+- The upload port is specified but the protocol of the selected port doesn't match any of the available upload protocols for a board
+
+Let's see some examples to clarify:
+
+```
+board1.upload.tool.no-port=openocd_without_port
+
+board2.upload.tool.serial=bossac
+board2.upload.tool.no-port=openocd_without_port
+
+board3.upload.tool.serial=bossac
+```
+
+In the `board1` case: the `openocd_without_port` recipe will be always used, even if a port has been selected by the user.
+
+In the `board2` case: the `bossac` recipe will be used if the port selected is a `serial` port, otherwise the `openocd_without_port` will be used in all other cases (even if a port has been selected by the user).
+
+In the `board3` case: the `bossac` recipe will be used if the port selected is a `serial` port, otherwise the upload will fail.
+
+A lot of existing platforms already have recipes without an explicit port address, in this case the upload tool specified in the old (non-pluggable) way will be considered as a `no-port` upload, for example let's consider the Arduino Zero board:
+
+```
+# Arduino Zero (Prorgamming Port)
+# ---------------------------------------
+arduino_zero_edbg.name=Arduino Zero (Programming Port)
+arduino_zero_edbg.vid.0=0x03eb
+arduino_zero_edbg.pid.0=0x2157
+
+arduino_zero_edbg.upload.tool=openocd        <--- CONSIDERED AS no-port PROT.
+arduino_zero_edbg.upload.protocol=sam-ba
+arduino_zero_edbg.upload.maximum_size=262144
+arduino_zero_edbg.upload.maximum_data_size=32768
+arduino_zero_edbg.upload.use_1200bps_touch=false
+arduino_zero_edbg.upload.wait_for_upload_port=false
+arduino_zero_edbg.upload.native_usb=false
+```
+
+in this case the upload definition will be always considered as a `no-port` by default, in other words, it will be automatically converted into:
```

Comment:
I was a bit confused by this at first, since it seems weird that you would use `no-port` for old upload tool definitions, since a lot of them will be serial instead. However, I then realized that defaulting to `no-port`, combined with the earlier definition that `no-port` is used when no matching protocol is available, effectively means that such a legacy protocol-less upload tool definition is used as a wildcard (i.e. used for *any* port, regardless of protocol). Maybe it would be good to make this explicit? i.e. not say it is used as a no-prot, but say that it should be used for *any* selected port, regardless of protocol. This makes it more obvious what the intention is, and decouples this intention from the "use no-port when no matching protocol is found"-policy.

Also, I think this wildcard policy is intended only for completely legacy board definitions (i.e. which specify no `upload.tool.<protocol>` entries), right? So once any protocol-specific entry is specified, the legacy entry will be completely ignored, right? The current wording might be read to mean that you might do:
```
board.upload.tool=openocd
board.upload.serial.tool=avrdude
```
and have it equivalent to:
```
board.upload.no-port.tool=openocd
board.upload.serial.tool=avrdude
```

but I think this is *not* the intention, right? If so, this should be made a bit more explicit maybe?

Reply:
> but I think this is not the intention, right? If so, this should be made a bit more explicit maybe?

correct.

You also got it right my intended use for the `no-port` in the legacy and I like your suggestions I'll try to rewrite it in a more explicit form.

Reply:
Ok while rewriting I found that `no-port` is really not much descriptive and confusing so I renamed it to `default`.
I also added a note about the backward compatibility helper that triggers only on non-pluggable platforms.